extensions/iam/distributed-identity/identity-did-web/src/main/java/org/eclipse/dataspaceconnector/iam/did/web/resolution/WebDidResolver.java

```diff
@@ -0,0 +1,133 @@
+    private static final String DID_SCHEME = "did";
+    private static final String DID_METHOD = "web";
+    private static final String DID_METHOD_PREFIX = DID_METHOD + ":";
+    private static final String HTTPS_PREFIX = " https://";
+    private static final String DID_DOCUMENT = "did.json";
+    private static final String WELL_KNOWN = "/.well-known/";
+
+    private OkHttpClient httpClient;
+    private ObjectMapper mapper;
+    private Monitor monitor;
+
+    /**
+     * Creates a resolver that executes DNS over HTTPs queries against the given DNS server.
+     */
+    public WebDidResolver(@Nullable URL dnsServer, OkHttpClient httpClient, ObjectMapper mapper, Monitor monitor) {
+        if (dnsServer != null) {
+            // use DNS over HTTPS for name lookups
+            var dns = new DnsOverHttps.Builder().client(httpClient).url(requireNonNull(HttpUrl.get(dnsServer))).includeIPv6(false).build();
+            this.httpClient = httpClient.newBuilder().dns(dns).build();
+        } else {
+            // use standard DNS lookups
+            this.httpClient = httpClient;
+        }
+        this.mapper = mapper;
+        this.monitor = monitor;
+    }
+
+    /**
+     * Creates a resolver that executes standard DNS lookups.
+     */
+    public WebDidResolver(OkHttpClient httpClient, ObjectMapper mapper, Monitor monitor) {
+        this(null, httpClient, mapper, monitor);
+    }
+
+    @Override
+    public @NotNull String getMethod() {
+        return DID_METHOD;
+    }
+
+    @Override
+    @Nullable
+    public DidDocument resolve(String didKey) {
+        try {
+            var request = new Request.Builder().url(keyToUrl(didKey)).get().build();
+
+            try (var response = httpClient.newCall(request).execute()) {
+                if (response.code() != 200) {
+                    monitor.severe(format("Error resolving DID: %s. HTTP Code was: %s", didKey, response.code()));
+                    return null;
+                }
+                try (var body = response.body()) {
+                    if (body == null) {
+                        monitor.severe("DID response contained an empty body: " + didKey);
+                        return null;
+                    }
+                    return mapper.readValue(body.string(), DidDocument.class);
+                }
+            } catch (IOException e) {
+                monitor.severe("Error resolving DID: " + didKey, e);
+                return null;
+            }
+        } catch (URISyntaxException | IllegalArgumentException | MalformedURLException e) {
+            monitor.severe("Invalid DID key: " + didKey, e);
+            return null;
+        }
+    }
+
+    /**
+     * Converts a DID URN to a URL according to the Web DID specification, .cf https://w3c-ccg.github.io/did-method-web/#read-resolve.
+     */
+    String keyToUrl(String didKey) throws URISyntaxException, IllegalArgumentException, MalformedURLException {
+        var uri = new URI(didKey);
```

Comment:
This operation could be encapsulated using a `Did` class, that could be passed to the resolver simplifying the validation process, in fact, the scheme validation is already done by `DidResolverRegistryImpl`

Reply:
I don't think it should. DID keys are serializable as strings (they are URNs) per the DID spec and this is the general DidResolver interface, which is used in various code paths. Hence, we can't assume the DID is valid and wrapping it in a class doesn't really do anything.